src/verify.c

```diff
@@ -0,0 +1,182 @@
+  BIO *bio_document = BIO_new_mem_buf(document, document_l);
+  BIO *bio_pkcs7 = BIO_new_mem_buf(pkcs7, pkcs7_l);
+
+  // The data structures needed for verification
+  PKCS7 *p7 = NULL;
+  X509_STORE *store = NULL;
+  STACK_OF(X509) *certs = NULL;
+  X509 *cert = NULL;
+
+  p7 = PEM_read_bio_PKCS7(bio_pkcs7, NULL, NULL, NULL);
+  if (p7 == NULL) {
+    rv = VF_EXCEPTION;
+    goto end;
+  }
+
+  store = X509_STORE_new();
+  if (store == NULL) {
+    rv = VF_EXCEPTION;
+    goto end;
+  }
+
+  certs = sk_X509_new_null();
+  if (certs == NULL) {
+    rv = VF_EXCEPTION;
+    goto end;
+  }
+
+  cert = PEM_read_bio_X509(bio_pubkey, NULL, NULL, NULL);
+  if (cert == NULL) {
+    rv = VF_EXCEPTION;
+    goto end;
+  }
+
+  if (0 == sk_X509_push(certs, cert)) {
+    rv = VF_EXCEPTION;
+    goto end;
+  }
+
+  // NOVERIFY is set to avoid validating the certificate chain for signing.
+  // Since the signatures this library is designed to verify will always be
+  // self-signed, the NOVERIFY option is required for the verification to work
+  if (1 == PKCS7_verify(p7, certs, store, bio_document, NULL,
+                        PKCS7_NOINTERN | PKCS7_NOVERIFY)) {
+    rv = VF_SUCCESS;
+  } else {
+    // The last error in the error queue ought to be a "signature verification"
+    // error as that's the last error to emit when a document validates to
+    // invalid.  In order to determine whether to treat this as VF_FAIL or
+    // VF_EXCEPTION, first determine whether this error is a signature
+    // verification error or not.  If it is, mark this invocation of VF_verify
+    // as VF_FAIL, then clear the error queue so that if future errors occur,
+    // handle them as exceptions.  This also ensures that exceptions during
+    // verification are handled differently to invalid signatures
+    unsigned long errorNum = ERR_peek_last_error();
+
+    if (ERR_GET_LIB(errorNum) == ERR_LIB_PKCS7 &&
+        ERR_GET_FUNC(errorNum) == PKCS7_F_PKCS7_VERIFY &&
+        ERR_GET_REASON(errorNum) == PKCS7_R_SIGNATURE_FAILURE) {
+      ERR_clear_error();
+      rv = VF_FAIL;
+    } else {
+      rv = VF_EXCEPTION;
+    }
+  }
+
+end:
+  if (!BIO_free(bio_document) || !BIO_free(bio_pubkey) ||
+      !BIO_free(bio_pkcs7)) {
+    rv = VF_EXCEPTION;
+  }
+
+  PKCS7_free(p7);
+  X509_STORE_free(store);
+  sk_X509_free(certs);
+  X509_free(cert);
+
+  struct Error *head = NULL;
+
+  unsigned long errorNum = ERR_peek_error();
+
+  if (!errorNum && rv == VF_EXCEPTION) {
+    // This case is for there being an exception signaled in this file but
+    // there isn't a corresponding error in the OpenSSL error queue.  Ideally,
+    // we'd use ERR_put_error to insert error messages which we could use to
+    // display using a single error reporting system.
+    head = malloc(sizeof(struct Error));
+    head->reason = "IID-Verify Exception";
+    head->lib = __FILE__;
+    head->func = "VF_verify";
+    head->next = NULL;
+    head->line = __LINE__;
```

Comment:
`head->file` is not set here, leaving it at an undefined value

Reply:
and correspondingly, `VF_err_fmt` omits the filename, so this never crashes anything (until someone notices and fixes that..)

Reply:
yep!  I've added that and now including that in the error messages.  That said, snprintf versions that I have been using do not crash when a `char*` is `NULL`:
```C
[jhford@jhford-desktop iid-verify]$ cat printf.c 
#include <stdio.h>
#include <stdlib.h>

int main(void) {
  struct E {
    char* d;
  };
  struct E *e = malloc(sizeof(struct E));
  e->d = NULL;

  char *msg = malloc(1001);
  snprintf(msg, 1000, "%s\n", e->d);
  printf("%s", msg);
  return 0;
}
[jhford@jhford-desktop iid-verify]$ clang -Wall -Werror printf.c -o test
[jhford@jhford-desktop iid-verify]$ ./test 
(null)
[jhford@jhford-desktop iid-verify]$ echo $?
0
[jhford@jhford-desktop iid-verify]$
```

Reply:
The code uses `malloc` to allocate the memory for the Error, so the `file` field contained whatever 64 bits were in that location already -- not necessarily NULL.  But now it's initialized so we're all good :)

Reply:
heh, yah, you're right.  I always forget that C does not initialize to 0s on malloc.